adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -834,59 +808,55 @@ protected final void onUnsubscribe(
             .withTag(TracingHelper.TAG_QOS.getKey(), qos.toString())
-            .withTag(TracingHelper.TAG_CLIENT_ID.getKey(), context.deviceEndpoint().clientIdentifier())
+            .withTag(TracingHelper.TAG_CLIENT_ID.getKey(), endpoint.clientIdentifier())
             .start();
-        context.setTimer(getMetrics().startTimer());
 
-        applyTenantTraceSamplingPriority(context, span)
-                .compose(v -> {
-                    context.setTracingContext(span.context());
-                    return checkTopic(context);
-                })
-                .compose(ok -> onPublishedMessage(context))
-                .onComplete(processing -> {
-                    if (processing.succeeded()) {
-                        Tags.HTTP_STATUS.set(span, HttpURLConnection.HTTP_ACCEPTED);
-                        onMessageSent(context);
-                    } else {
-                        if (processing.cause() instanceof ServiceInvocationException) {
-                            final ServiceInvocationException sie = (ServiceInvocationException) processing.cause();
-                            Tags.HTTP_STATUS.set(span, sie.getErrorCode());
-                        } else {
-                            Tags.HTTP_STATUS.set(span, HttpURLConnection.HTTP_INTERNAL_ERROR);
-                        }
-                        if (processing.cause() instanceof ClientErrorException) {
-                            // nothing to do
-                        } else {
-                            onMessageUndeliverable(context);
-                        }
-                        TracingHelper.logError(span, processing.cause());
-                        if (context.deviceEndpoint().isConnected()) {
-                            span.log("closing connection to device");
-                            context.deviceEndpoint().close();
-                        }
+        MqttContext.fromPublishPacket(message, endpoint, tenantAndAuthIdExtractor, authenticatedDevice, span.context())
+                .compose(context -> {
+                    context.setTimer(getMetrics().startTimer());
+                    applyTenantTracingOptions(context.tenantObject(), context.authId(), span, context);
+
+                    return onPublishedMessage(context)
+                            .recover(thr -> {
+                                if (thr instanceof ClientErrorException) {
+                                    // nothing to do
+                                } else {
+                                    onMessageUndeliverable(context);
+                                }
+                                return Future.failedFuture(thr);
+                            }).map(ok -> {
+                                Tags.HTTP_STATUS.set(span, HttpURLConnection.HTTP_ACCEPTED);
+                                onMessageSent(context);
+                                return null;
+                            });
+                }).recover(thr -> {
```

Comment:
Similar to the above: There's an `onComplete` coming afterwards which has to run after this `recover` case. Therefore I'd rather use composed Futures instead of one Future with multiple handlers.